src/mosquitto_stubs.c

```diff
@@ -217,96 +255,58 @@ CAMLprim value mqtt_publish(value mqtt, value msg) {
   memcpy(payload, String_val(Field(msg, 2)), payload_len);
 
   caml_release_runtime_system();
-
   rc = mosquitto_publish(mq->conn, mid_pt, topic, payload_len, payload, qos, retain);
+  int lerrno = errno;
+  caml_acquire_runtime_system();
+  result = wrap_result(rc, Val_unit, lerrno);
   free(topic);
   free(payload);
 
-  caml_acquire_runtime_system();
-
-  if (MOSQ_ERR_SUCCESS == rc) {
-    result = RESULT_OK;
-    Store_field(result, 0, Val_unit);
-  } else {
-    switch (rc) {
-      case MOSQ_ERR_INVAL: lerrno = EINVAL; break;
-      case MOSQ_ERR_NOMEM: lerrno = ENOMEM; break;
-      case MOSQ_ERR_NO_CONN: lerrno = ENOTCONN; break;
-      case MOSQ_ERR_PROTOCOL: lerrno = EPROTOTYPE; break;
-      case MOSQ_ERR_PAYLOAD_SIZE: lerrno = EMSGSIZE; break;
-      case MOSQ_ERR_ERRNO: lerrno = errno; break;
-      default: assert(false);
-    }
-    perrno = caml_alloc(2, 0);
-    Store_field(perrno, 0, eunix); // `EUnix
-    Store_field(perrno, 1, unix_error_of_code(lerrno));
-
-    result = RESULT_ERROR;
-    Store_field(result, 0, perrno);
-  }
-
   CAMLreturn(result);
 }
 
 CAMLprim value mqtt_subscribe(value mqtt, value topic, value qos) {
   CAMLparam3(mqtt, topic, qos);
-  CAMLlocal2(perrno, result);
+  CAMLlocal1(result);
 
   char *topic_;
   size_t topic_len;
   struct ocmq *mq;
-  int rc, lerrno, qos_;
+  int rc, qos_;
 
   qos_ = Long_val(qos);
-  mq = (struct ocmq*)mqtt;
+  mq = *(OCMQ(mqtt));
+  if (mq->conn == NULL) caml_invalid_argument("Mosquitto: already destroyed");
 
   topic_len = caml_string_length(topic);
   topic_ = calloc(1, topic_len + 1);
   memcpy(topic_, String_val(topic), topic_len);
   topic_[topic_len] = '\0';
 
   caml_release_runtime_system();
-
   rc = mosquitto_subscribe(mq->conn, NULL, topic_, qos_);
-  free(topic_);
-
+  int lerrno = errno;
   caml_acquire_runtime_system();
-
-  if (MOSQ_ERR_SUCCESS == rc) {
-    result = RESULT_OK;
-    Store_field(result, 0, Val_unit);
-  } else {
-    switch (rc) {
-      case MOSQ_ERR_INVAL: lerrno = EINVAL; break;
-      case MOSQ_ERR_NOMEM: lerrno = ENOMEM; break;
-      case MOSQ_ERR_NO_CONN: lerrno = ENOTCONN; break;
-      default: assert(false);
-    }
-    perrno = caml_alloc(2, 0);
-    Store_field(perrno, 0, eunix); // `EUnix
-    Store_field(perrno, 1, unix_error_of_code(lerrno));
-
-    result = RESULT_ERROR;
-    Store_field(result, 0, perrno);
-  }
+  result = wrap_result(rc, Val_unit, lerrno);
+  free(topic_);
 
   CAMLreturn(result);
 }
 
 void mqtt_callback_msg(struct mosquitto *m, void *obj, const struct mosquitto_message *msg_) {
-  static value * f = NULL;
   struct ocmq *mq;
   size_t topic_len;
   value msg, payload;
 
   mq = (struct ocmq*)obj;
+  if (mq->conn == NULL) return;
 
   caml_acquire_runtime_system();
 
-  if (NULL == f) {
-    f = caml_named_value(mq->uid[CBMESSAGE]);
+  if (NULL == mq->cb[CBMESSAGE]) {
+    mq->cb[CBMESSAGE] = caml_named_value(mq->uid[CBMESSAGE]);
```

Comment:
Removed the attempt of an optimization, as it cannot work when there are several handles open.